firebase-ml-modeldownloader/src/main/java/com/google/firebase/ml/modeldownloader/FirebaseModelDownloader.java

```diff
@@ -95,21 +112,77 @@ public static FirebaseModelDownloader getInstance(@NonNull FirebaseApp app) {
   public Task<CustomModel> getModel(
       @NonNull String modelName,
       @NonNull DownloadType downloadType,
-      @Nullable CustomModelDownloadConditions conditions) {
+      @Nullable CustomModelDownloadConditions conditions)
+      throws Exception {
+    CustomModel localModel = sharedPreferencesUtil.getCustomModelDetails(modelName);
+    switch (downloadType) {
+      case LOCAL_MODEL:
+        if (localModel != null) {
+          return Tasks.forResult(localModel);
+        }
+        Task<CustomModel> modelDetails =
+            modelDownloadService.getCustomModelDetails(
+                firebaseOptions.getProjectId(), modelName, null);
+
+        // no local model - start download.
+        return modelDetails.continueWithTask(
+            executor,
+            modelDetailTask -> {
+              if (modelDetailTask.isSuccessful()) {
+                // start download
+                return fileDownloadService
+                    .download(modelDetailTask.getResult(), conditions)
+                    .continueWithTask(
+                        executor,
+                        downloadTask -> {
+                          if (downloadTask.isSuccessful()) {
+                            // read the updated model
+                            CustomModel downloadedModel =
+                                sharedPreferencesUtil.getCustomModelDetails(modelName);
+                            // TODO(annz) trigger file move here as well... right now it's temp
+                            // call loadNewlyDownloadedModelFile
+                            return Tasks.forResult(downloadedModel);
+                          }
+                          return Tasks.forException(new Exception("File download failed."));
+                        });
+              }
+              return Tasks.forException(modelDetailTask.getException());
+            });
+      case LATEST_MODEL:
+        // check for latest model and download newest
+        break;
+      case LOCAL_MODEL_UPDATE_IN_BACKGROUND:
+        // start download in back ground return current model if not null.
+        break;
+    }
     throw new UnsupportedOperationException("Not yet implemented.");
   }
 
-  /** @return The set of all models that are downloaded to this device. */
+  /**
+   * Triggers the move to permanent storage of successful model downloads and list all models
+   * downloaded to device.
+   *
+   * @return The set of all models that are downloaded to this device, triggers completion of file
+   *     moves for completed model downloads.
+   */
   @NonNull
   public Task<Set<CustomModel>> listDownloadedModels() {
+    // trigger completion of file moves for download files.
+    try {
+      fileDownloadService.maybeCheckDownloadingComplete();
+    } catch (Exception ex) {
+      System.out.println("Error checking for in progress downloads: " + ex.getMessage());
```

Comment:
Log maybe?

Reply:
Yes - changing all these to logs in the next major PR. Just started working on it this morning.

Reply:
Happy to change it earlier if you want - was just trying to keep this change as small as possible.

Reply:
That's OK, you can do it later.